### PR TITLE
fixes mm-1351 adds local server storage

### PR DESCRIPTION
**Pull request by jwilander**
Edited the config file

### CODE REVIEW

File: config/config_docker.json [hunk 1]
```diff
@@ -20,7 +20,9 @@
         "InviteSalt": "gxHVDcKUyP2y1eiyW8S8na1UYQAfq6J6",
         "PublicLinkSalt": "TO3pTyXIZzwHiwyZgGql7lM7DG3zeId4",
         "ResetSalt": "IPxFzSfnDFsNsRafZxz8NaYqFKhf9y2t",
-        "AnalyticsUrl": ""
+        "AnalyticsUrl": "",
+        "UseLocalStorage": true,
+        "StorageDirectory": "/mattermost/"
```
Comment (coreyhulen) [inline config/config_docker.json]: does the runas user in docker have permissions to create dirs in this location?

File: config/config.json [hunk 2]
```diff
@@ -20,7 +20,9 @@
         "InviteSalt": "gxHVDcKUyP2y1eiyW8S8na1UYQAfq6J6",
         "PublicLinkSalt": "TO3pTyXIZzwHiwyZgGql7lM7DG3zeId4",
         "ResetSalt": "IPxFzSfnDFsNsRafZxz8NaYqFKhf9y2t",
-        "AnalyticsUrl": ""
+        "AnalyticsUrl": "",
+        "UseLocalStorage": false,
+        "StorageDirectory": ""
```
Comment (coreyhulen) [inline config/config.json]: maybe turn on and set to relative path of data/ for unit testing and local dev builds?  (0/5)  then add data/ to the gitignore

File: config/config_docker.json (hunk 1 above)
Comment (coreyhulen) [inline config/config_docker.json]: maybe /mattermost/data to sit along side bin and config folders?  0/5

now that I think about it I'm not sure what dir docker installers our app in.

Comment (jwilander) [inline config/config_docker.json]: Docker installs us in /go/src/github.com/mattermost/platform and yeah /mattermost/data might be better as it gives a better indication as to what's being stored

File: config/config.json (hunk 2 above)
Comment (jwilander) [inline config/config.json]: Yup, good idea

**Comment by jwilander** [comment]
Ready for re-review

**Comment by jwilander** [comment]
Ready for re-review

File: api/file.go [hunk 3]
```diff
@@ -94,28 +89,23 @@ func uploadFile(c *Context, w http.ResponseWriter, r *http.Request) {
 		buf := bytes.NewBuffer(nil)
 		io.Copy(buf, file)
 
-		ext := filepath.Ext(files[i].Filename)
-
 		uid := model.NewId()
 
 		path := "teams/" + c.Session.TeamId + "/channels/" + channelId + "/users/" + c.Session.UserId + "/" + uid + "/" + files[i].Filename
```
Comment (jwilander) [inline api/file.go]: Good catch, I'll add sanitization to prevent relative paths and add a unit test for it

File: Dockerfile [hunk 4]
```diff
@@ -88,12 +88,15 @@ ADD . /go/src/github.com/mattermost/platform
...
+RUN cd /go/src/github.com/mattermost/platform; godep restore
 RUN go install github.com/mattermost/platform
-RUN cd /go/src/github.com/mattermost/platform/web/react; npm install 
+RUN cd /go/src/github.com/mattermost/platform/web/react; npm install
 
 RUN chmod +x /go/src/github.com/mattermost/platform/docker/docker-entry.sh
 ENTRYPOINT /go/src/github.com/mattermost/platform/docker/docker-entry.sh
 
+# Create default storage directory
+RUN mkdir /mattermost/
```
Comment (coreyhulen) [inline Dockerfile]: at root?

File: utils/urlencode.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+// See License.txt for license information.
+
+package utils
+
+import (
+	"net/url"
+	"strings"
+)
+
+func UrlEncode(str string) string {
```
Comment (coreyhulen) [inline utils/urlencode.go]: might look at the diffs with stuff in html and html/templatepkgs (0/5)

func HTMLEscapeString(s string) string
func EscapeString(s string) string

File: Dockerfile (hunk 4 above)
Comment (jwilander) [inline Dockerfile]: Yeah, that matches what's in the docker config but I'm 0/5 on where we put it

Comment (jwilander) [inline Dockerfile]: The Dockerfile doesn't actually need this line but I added to be explicit. The server will automatically create the directories it needs to save files

File: utils/urlencode.go (hunk 5 above)
Comment (jwilander) [inline utils/urlencode.go]: Those methods are for HTML escaping, aren't they? This is for URL encoding which is different. I also searched to see if an existing url encoder function existed in a go lib but couldn't find one

File: Dockerfile (hunk 4 above)
Comment (coreyhulen) [inline Dockerfile]: na..it's fine.  just wasn't sure.

File: utils/urlencode.go (hunk 5 above)
Comment (jwilander) [inline utils/urlencode.go]: Also the url.QueryEscape function works perfectly for this except it encodes spaces to "+" which gets all confused in the URL when there are any real "+" in the filename

**Pull request merged**